src/puppetlabs/puppetdb/admin.clj

```diff
@@ -0,0 +1,56 @@
+(ns puppetlabs.puppetdb.admin
+  (:require [clojure.java.io :as io]
+            [clojure.tools.logging :as log]
+            [compojure.core :as compojure]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [puppetlabs.puppetdb.archive :as archive]
+            [puppetlabs.puppetdb.cheshire :as json]
+            [puppetlabs.puppetdb.cli.export :as export]
+            [puppetlabs.puppetdb.cli.import :as import]
+            [puppetlabs.puppetdb.http :as http]
+            [puppetlabs.puppetdb.middleware :as mid]
+            [puppetlabs.trapperkeeper.core :as trapperkeeper]
+            [ring.middleware.multipart-params :as mp]
+            [clj-time.core :refer [now]]
+            [ring.util.io :as rio]))
+
+(def query-api-version :v4)
+
+(defn munge-nodes-data
+  [nodes]
+  (as-> nodes $
+    (select-keys $ [:certname :facts_timestamp :catalog_timestamp :report_timestamp])
+    (kitchensink/mapvals str $)))
+
+(defn export-app
+  [buffer query-fn]
+  (export/export! buffer
+                  (->> (query-fn :nodes query-api-version nil nil doall)
+                       (map munge-nodes-data)
+                       (mapcat (partial export/get-node-data query-fn)))))
+
+(defn build-app
+  [submit-command-fn query-fn {:keys [authorizer]}]
+  (-> (compojure/routes
+       (mp/wrap-multipart-params
+        (compojure/POST "/v1/archive" request
+                        (let [{{:strs [archive command_versions]} :multipart-params} request]
+                          (import/import! (:tempfile archive)
+                                          (json/parse-string command_versions true)
+                                          submit-command-fn)
+                            (http/json-response {:ok true}))))
+       (compojure/GET "/v1/archive" []
+                      (http/streamed-tar-response #(export-app % query-fn)
+                                                  (format "puppetdb-export-%s.tgz" (now)))))
+      (mid/wrap-with-puppetdb-middleware authorizer)))
+
+(trapperkeeper/defservice admin-service
+  [[:PuppetDBServer shared-globals query]
+   [:PuppetDBCommand submit-command]
+   [:WebroutingService add-ring-handler get-route]]
+
+  (start [this context]
+         (->> (build-app submit-command query (shared-globals))
+              (compojure/context (get-route this) [])
+              (add-ring-handler this))
+         context))
```

Comment:
we have the same problem with 404s on this admin endpoint as we did with PDB-1509. I don't mind having a separate ticket to fix that at the middleware level, but we should track that work